### ISSUE TITLE
[k3s-upgrade] Upgrade-pods fail at the moment of killing the k3s process

### ISSUE THREAD

**Issue by ChristianCiach** [k3s-io/k3s#11347]
**Environmental Info:**
K3s Version: `v1.31.2+k3s1` (but doesn't really matter)

Node(s) CPU architecture, OS, and Version: `Linux k3s-server-1 5.14.0-427.28.1.el9_4.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Aug 2 03:44:10 EDT 2024 x86_64 x86_64 x86_64 GNU/Linux` (... but shouldn't matter)

Cluster Configuration:

3 servers and 1 agent as AlmaLinux 9 VMs (I could share my Vagrant file for reproduction).

**Describe the bug:**
Every time the the system-upgrade-controller executes a `Job` to upgrade a node using the `k3s-upgrade` container, the job's pod fails just at the moment of killing the old `k3s` process, probably because the control-plane loses connection to the kubelet of the node being upgraded. But the upgrade-script completes successfully:

```
+ info 'Deploying new k3s binary to /usr/local/bin/k3s'
+ echo '[INFO] ' 'Deploying new k3s binary to /usr/local/bin/k3s'
+ cp /opt/k3s /host/usr/local/bin/k3s
[INFO]  Deploying new k3s binary to /usr/local/bin/k3s
[INFO]  Restoring k3s bin context
+ '[' -n system_u:object_r:bin_t:s0 ]
+ info 'Restoring k3s bin context'
+ echo '[INFO] ' 'Restoring k3s bin context'
+ setfilecon system_u:object_r:bin_t:s0 /host/usr/local/bin/k3s
[INFO]  K3s binary has been replaced successfully
+ info 'K3s binary has been replaced successfully'
+ echo '[INFO] ' 'K3s binary has been replaced successfully'
+ return
+ kill_k3s_process
+ kill -SIGTERM 19484
+ info 'Successfully killed old k3s pid 19484'
[INFO]  Successfully killed old k3s pid 19484
+ echo '[INFO] ' 'Successfully killed old k3s pid 19484'
Stream closed EOF for system-upgrade/apply-server-plan-on-k3s-server-1-with-0ad8f12c7e9b8fa243-sndx7 (upgrade)
```

This leads to every upgrade-job spawning two pods, because the first pod of each job fails with status `Unknown`:

```
# kubectl get pods -o wide -n system-upgrade
NAME                                                              READY   STATUS      RESTARTS   AGE     IP               NODE           NOMINATED NODE   READINESS GATES
apply-agent-plan-on-k3s-agent-1-with-0ad8f12c7e9b8fa24394-hc97h   0/1     Unknown     0          3m33s   192.168.122.21   k3s-agent-1    <none>           <none>
apply-agent-plan-on-k3s-agent-1-with-0ad8f12c7e9b8fa24394-rs4fp   0/1     Completed   0          101s    192.168.122.21   k3s-agent-1    <none>           <none>
apply-server-plan-on-k3s-server-1-with-0ad8f12c7e9b8fa243-8zv9r   0/1     Completed   0          10m     192.168.122.11   k3s-server-1   <none>           <none>
apply-server-plan-on-k3s-server-1-with-0ad8f12c7e9b8fa243-sndx7   0/1     Unknown     0          13m     192.168.122.11   k3s-server-1   <none>           <none>
apply-server-plan-on-k3s-server-2-with-0ad8f12c7e9b8fa243-cmthr   0/1     Completed   0          8m      192.168.122.12   k3s-server-2   <none>           <none>
apply-server-plan-on-k3s-server-2-with-0ad8f12c7e9b8fa243-vsn2s   0/1     Unknown     0          10m     192.168.122.12   k3s-server-2   <none>           <none>
apply-server-plan-on-k3s-server-3-with-0ad8f12c7e9b8fa243-6xqmf   0/1     Unknown     0          7m34s   192.168.122.13   k3s-server-3   <none>           <none>
apply-server-plan-on-k3s-server-3-with-0ad8f12c7e9b8fa243-x7m8x   0/1     Completed   0          6m33s   192.168.122.13   k3s-server-3   <none>           <none>
system-upgrade-controller-b599d7b64-gmlbb                         1/1     Running     0          19m     10.42.2.3        k3s-server-3   <none>           <none>
```

When asking Kubernetes to describe the failed Pods, these are the last events:

```
  Warning  FailedMount       3m1s   kubelet            MountVolume.SetUp failed for volume "secret-inventories-kd8m7bggc7" : object "system-upgrade"/"inventories-kd8m7bggc7" not registered
  Warning  FailedMount       3m1s   kubelet            MountVolume.SetUp failed for volume "secret-scripts-dgcf5g87g5" : object "system-upgrade"/"scripts-dgcf5g87g5" not registered
  Warning  FailedMount       3m1s   kubelet            MountVolume.SetUp failed for volume "kube-api-access-mspvx" : object "system-upgrade"/"kube-root-ca.crt" not registered
```

The first two mounts are used by our `prepare` container to deploy the config file `/etc/rancher/k3s/config.yaml`, but the same happens without our custom `prepare` step.

The second pod of a job succeeds because the binary has been successfully replaced by the first pod, so the second pod doesn't kill the process:

```
+ info 'Comparing old and new binaries'
+ echo '[INFO] ' 'Comparing old and new binaries'
+ sha256sum /opt/k3s /host/usr/local/bin/k3s
+ BIN_CHECKSUMS='256ca852ff2ca165afbc14ed34df450aa1295e4eb01e3d1da9f251f37e546712  /opt/k3s
256ca852ff2ca165afbc14ed34df450aa1295e4eb01e3d1da9f251f37e546712  /host/usr/local/bin/k3s'
+ '[' 0 '!=' 0 ]
+ echo '256ca852ff2ca165afbc14ed34df450aa1295e4eb01e3d1da9f251f37e546712  /opt/k3s
256ca852ff2ca165afbc14ed34df450aa1295e4eb01e3d1da9f251f37e546712  /host/usr/local/bin/k3s'
+ awk '{print $1}'
+ uniq
+ wc -l
+ BIN_COUNT=1
+ '[' 1 '=' 1 ]
+ info 'Binary already been replaced'
+ echo '[INFO] ' 'Binary already been replaced'
+ exit 0
[INFO]  Binary already been replaced
Stream closed EOF for system-upgrade/apply-server-plan-on-k3s-server-1-with-0ad8f12c7e9b8fa243-8zv9r (upgrade)
```

Since the second pod of each job succeeds and the upgrade plans (including our custom `prepare` container) are idempotent, this doesn't prevent us from upgrading the nodes. I mostly report this to understand if this behaves as designed.

**Steps To Reproduce:**
Really nothing special. Just install a vanilla K3s cluster using the official installation scripts, deploy the system-upgrade-controller and deploy an upgrade plan that uses the `k3s-upgrade` images:

```yaml
apiVersion: upgrade.cattle.io/v1
kind: Plan
metadata:
  name: server-plan
  namespace: system-upgrade
spec:
  concurrency: 1
  cordon: true
  nodeSelector:
    matchExpressions:
    - key: node-role.kubernetes.io/control-plane
      operator: In
      values:
      - "true"
  prepare:
    command:
    - bash
    - -ec
    - |
      cp /tmp/scripts/run.sh /tmp/run.sh
      chmod +x /tmp/run.sh
      /tmp/run.sh
    envs:
    - name: NODE_NAME
      valueFrom:
        fieldRef:
          fieldPath: spec.nodeName
    image: company.registry.local/platform/k3s-deployment:k3s-auto-upgrade-ng-dev
  secrets:
  - name: inventories-kd8m7bggc7
    path: /tmp/inventories
  - ignoreUpdates: true
    name: scripts-dgcf5g87g5
    path: /tmp/scripts
  serviceAccountName: system-upgrade
  tolerations:
  - operator: Exists
  upgrade:
    image: docker.io/rancher/k3s-upgrade:v1.31.2-k3s1
  version: v1.31.2+k3s1
```

This happens with both k3s servers and agents.

**Expected behavior:**
I didn't expect every node-specific upgrade job to spawn two pods where the first one always fails. But maybe this is actually working as designed?

**Actual behavior:**
Failing pods that lead to duplicated upgrade pods.

**Comment by ChristianCiach** [comment]
I have no idea how these failing pods could possibly be avoided. Maybe `k3s-upgrade` could slightly delay the killing of the pod and disown the script that kills the process? This would probably give the pod enough time to properly complete, but then we wouldn't know if the killing of the process fails.

**Bonus question**

Please let me know if I should create a separate issue for this or if it's working as expected:

It seems that k3s must be stopped before executing `k3s certificate rotate`. I first tried to do the following while k3s is still running: `k3s certificate rotate && systemctl restart k3s`, but then k3s doesn't start because of missing certificate files (until I restart k3s *again*). So it seems to be important to stop k3s beforehand: `systemctl stop k3s && k3s certificate rotate && systemctl start`.

I am asking this here because I tried to rotate the certificates as part of the `prepare` step of the upgrade-plan and let the `k3s-upgrade` restart (by killing) the process. But it looks like I first need to stop the process, rotate the certificates and restart the process as part of the prepare-step, just so that k3s-upgrade can kill it again. That's fine, but I would prepare to restart the process only once per node.

**Comment by ChristianCiach** [comment]
**Another bonus question**

`k3s certificate rotate --help` claims to use `/etc/rancher/k3s/config.yaml` by default:

```
--config FILE, -c FILE      (config) Load configuration from FILE (default: "/etc/rancher/k3s/config.yaml") [$K3S_CONFIG_FILE]
```

Our `config.yaml` contains a `data-dir` value, but this gets ignored by `k3s certificate rotate`, so I need to pass `--data-dir` explicitly. Is this expected?

**Comment by brandond** [comment]
> The second pod of a job succeeds because the binary has been successfully replaced by the first pod, so the second pod doesn't kill the process:

Yes, this is how the k3s-upgrade image is intended to work.

**Comment by Ch4s3r** [comment]
Had the same issue with one of the two pods `apply-server-plan-on-k3s-with-a81b291ab92e98978bc5df23e5e-5rhpn` being stuck in `Unknown` status and the other one had the status `Completed`.
After 16 minutes both pods vanished.
